### ISSUE TITLE
Create new TBWrapper model

### ISSUE THREAD

**Issue by JeffAbrahamson** [transport-nantes/tn_web#912]
This should be relatively simple and quick...

What we want is something that lets you choose a TBObject type (dropdown), a published slug for that type (dropdown or type-ahead (eventually) but can be copy/paste by user in first iteration), and then prompts for open graph data (but not twitter).  And, of course, it needs a slug, which we should auto-generate from the provided slug, for example `underlying_slug + sha1(underlying_slug.encode()).hex()[:6]`.

On display (that is, when someone clicks `/tb/w/<slug>`) we look up the wrapper object, recover the underlying TBObject type and slug name, and display that _except_ (and this is the reason it exists) we pretend that that object has the open graph data from the wrapper object (and no twitter card data).

The challenge is to do this with a minimum of code changes and a maximum of simplicity, so let's think about it and design it before writing it.

The reason we want this is to permit re-sharing things we've previously shared where the only thing we want to change is the social media presentation.

_Aside: I'm increasingly inclined to deprecated twitter card data.  We typically upload the same info to both or else upload OG and let twitter fall back to OG.  But that's for another day._

**Comment by Shriukan33** [comment]
What if we override the set_social_context here ? 

https://github.com/transport-nantes/tn_web/blob/22ac17fb686d1fa4c7ac18cdd0fede7d9c11d50f/transport_nantes/topicblog/models.py#L452-L467

Also this would mean that we cannot have the same slug used say in tbitem and tbemail.
Currently, there isn't a unique constraint that would enforce that, we could still ignore it by setting a priority order (like if it finds it first on tbitem, then it returns that tbitem), but that could lead to unexpected behavior if this isn't explicit on the front end.

Could we set a unique constraint that spans multiple tables ?

**Comment by JeffAbrahamson** [comment]
> Also this would mean that we cannot have the same slug used say in tbitem and tbemail.

That's a constraint that (1) users would have trouble finding reasonable and (2) is going to be fiendishly error-prone to enforce.

**Comment by Shriukan33** [comment]
I totally missed that you suggested to add a dropdown on that. 
Then yeah, we could just fetch its type there and use the corresponding manager to get the desired object. 

I'd inherit TBWrapper from TBSocialBase, and override the set_social_context output with something like 

_views.py_
```python
class TBWrapperView(TopicBlogBaseView):

    def get_context_data(self, **kwargs):
        context = super().get_context_data(self, **kwargs)
        # We change the context here) 
        context["social"] = our_wrapper_set_social_context_function(context)
```

We also need to define self.model with a `get_tb_object_model()` method for TopicBlogBaseView's get_context_data to work.

**Comment by JeffAbrahamson** [comment]
Yeah, that sounds much better! 😍 

So once you have fetched the wrapper object and then fetched the underlying object, you can update the OG data, and you know the template to use.

Can you now render the template without repeating more than a couple lines of code?  If so, I think we're good to go.

The trick may be that some of the object fetching happens a bit too low-level in TBObject derived classes, and you want to separate fetch and render, I think?

**Comment by Shriukan33** [comment]
I'll write down the django process so it's clearer in my head :

GET -> /tb/w/the-slug
url path -> TBWrapperView.as_view()
TBWrapperView.get() has severals tasks : 
* fetch the wrapper thanks to its slug (UNIQUE constraint) and extract the manager to use from there (set with dropdown when created). We would reverse search the class using `apps.get_model("topicblog", "<the_value_in_dropdown>")` and save it as `self.model`
* Get a first context, which is defined a first time using super().get_context_data(), which is inherited from TopicBlogBaseView. We now have self.template_name, the tb_object, and default social context.
* Now we can override context["social"] with the values we have stored in the TBWrapper
* return and render

```python

class TopicBlogWrapperView(TopicBlogBaseView):

    def get_model(self, **kwargs):
        tb_wrapper = TBWrapperModel.objects.get_object_or_404(slug=self.kwargs["the_slug"])
        self.model = apps.get_model("topicblog", tb_wrapper.original_model)

    def override_social_context(context):
        social = {}
        social['twitter_title'] = self.twitter_title
        social['twitter_description'] = self.twitter_description
        social['twitter_image'] = self.twitter_image

        social['og_title'] = self.og_title
        social['og_description'] = self.og_description
        social['og_image'] = self.og_image

        context['social'] = social

    return context

    def get_context_data(self, **kwargs): 
        self.get_model(**kwargs)
        context = super().get_context_data(**kwargs)
        context = self.override_social_context(context)

    return context
```

Maybe it lacks some details but that's how I consider doing it at the moment. 

What do you mean by too low level in TB object derived classes ? What use case are you thinking of ?

**Comment by JeffAbrahamson** [comment]
Yes, I think that should work.  Nothing jumps out at me as causing problems.

When I read the TB code, I notice that fetch and render are often in the same function.  But it looks like you cleverly avoid that being a problem here.
